File: Data Structures & Algorithms/next-greater-element-i/submission-0.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        st, res, out = Stack(), {}, []
        
        for i in range(len(nums2)):
            top = st.top()
            while top != -1 and nums2[top] < nums2[i]:
                poped = st.pop()
                res[nums2[poped]] = nums2[i]
                top = st.top()
            st.push(i)

        return [res.get(n, -1) for n in nums1]

    

class Stack:
    def __init__(self):
        self.data = []
    
    def top(self):
        if not self.data:
            return -1
        return self.data[-1]
    
    def pop(self):
        if not self.data:
            return -1
        return self.data.pop()
    
    def push(self, index):
        self.data.append(index)

File: Data Structures & Algorithms/daily-temperatures/submission-0.go
func dailyTemperatures(t []int) []int {
    stack := Stack{[][]int{}}
    res := make([]int, len(t), len(t))

    for i, n := range t {
        if len(stack.data) == 0 {
            stack.Push(n, i)
        }


        for len(stack.data) != 0  {
            topNum, topIndex := stack.Top()
            if n <= topNum {
                break
            }
            res[topIndex] = i - topIndex
            stack.Pop()
        }

        stack.Push(n, i)
    }
    fmt.Println(stack)
    return res
}

type Stack struct {
    data [][]int
}

func (s *Stack) Pop() {
    s.data = s.data[:len(s.data)-1]
}

func (s *Stack) Top() (int, int) {
    ans := s.data[len(s.data)-1]
    return ans[0], ans[1]
}

func (s *Stack)  Push(n, i int) {
    s.data = append(s.data, []int{n, i})
}